pkg/pgmodel/delete.go

```diff
@@ -0,0 +1,95 @@
+package pgmodel
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const (
+	queryDeleteSeries = "select _prom_catalog.delete_series_from_metric($1, $2)"
+)
+
+type DBDeleter struct {
+	conn *pgxpool.Pool
+}
+
+// NewPgDeleter returns a series deleter that is dependent on the provided label set.
+func NewPgxDeleter(conn *pgxpool.Pool) *DBDeleter {
+	return &DBDeleter{conn}
+}
+
+// PgDelete
+type PgDelete struct {
+	conn *pgxpool.Pool
+}
+
+// PgDeleter returns a new PgDelete that can be used to delete series. It is goroutine safe.
+func (pgDelr *DBDeleter) PgDeleter() *PgDelete {
+	return &PgDelete{pgDelr.conn}
+}
+
+// getMetricNameSeriesIDFromMatchers returns the metric name list and the corresponding series ID array
+// as a matrix.
+func (pgDel *PgDelete) getMetricNameSeriesIDFromMatchers(matchers []*labels.Matcher) ([]string, [][]SeriesID, error) {
+	_, clauses, values, err := buildSubQueries(matchers)
+	if err != nil {
+		return nil, nil, fmt.Errorf("delete series from matchers: %w", err)
+	}
+	query := buildMetricNameSeriesIDQuery(clauses)
+	rows, err := pgDel.conn.Query(context.Background(), query, values...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("build metric name series: %w", err)
+	}
+	metricNames, correspondingSeriesIDs, err := getSeriesPerMetric(rows)
+	if err != nil {
+		return nil, nil, fmt.Errorf("series per metric: %w", err)
+	}
+	return metricNames, correspondingSeriesIDs, nil
+}
+
+// DeleteSeries deletes the series that matches the provided label_matchers.
+func (pgDel *PgDelete) DeleteSeries(matchers []*labels.Matcher) ([]string, []SeriesID, int, error) {
+	var (
+		deletedSeriesIDs   []SeriesID
+		totalChunksTouched int
+		err                error
+		metricsTouched     = make(map[string]struct{})
+	)
+	metricNames, seriesIDMatrix, err := pgDel.getMetricNameSeriesIDFromMatchers(matchers)
+	if err != nil {
+		return nil, nil, -1, fmt.Errorf("delete-series: %w", err)
+	}
+	for metricIndex, metricName := range metricNames {
+		seriesIDs := seriesIDMatrix[metricIndex]
+		for _, seriesID := range seriesIDs {
```

Comment:
As long as the second parameter is metric name, the caller need to ensure that the seriesIDs belong to that metric only, else it would be clumsy. Maybe, getting rid of it solves the matter here?